src/table.ts

```diff
@@ -778,137 +752,79 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
-        // Readjust and/or remove ranges based on previous valid row reads.
-        // Iterate backward since items may need to be removed.
-        for (let index = ranges.length - 1; index >= 0; index--) {
-          const range = ranges[index];
-          const startValue = is.object(range.start)
-            ? (range.start as BoundData).value
-            : range.start;
-          const endValue = is.object(range.end)
-            ? (range.end as BoundData).value
-            : range.end;
-          const startKeyIsRead =
-            !startValue ||
-            TableUtils.lessThanOrEqualTo(
-              startValue as string,
-              lastRowKey as string
-            );
-          const endKeyIsNotRead =
-            !endValue ||
-            (endValue as Buffer).length === 0 ||
-            TableUtils.lessThan(lastRowKey as string, endValue as string);
-          if (startKeyIsRead) {
-            if (endKeyIsNotRead) {
-              // EndKey is not read, reset the range to start from lastRowKey open
-              range.start = {
-                value: lastRowKey,
-                inclusive: false,
-              };
-            } else {
-              // EndKey is read, remove this range
-              ranges.splice(index, 1);
-            }
-          }
+    (() => {
+      /*
+      This end function was redefined in a fix to a data loss issue with streams
+      to allow the user to cancel the stream and instantly stop receiving more
+      data in the stream.
+       */
+      userStream.end = (
+        chunkOrCb: () => void | Row,
+        encodingOrCb?: BufferEncoding | (() => void),
+        cb?: () => void
+      ) => {
+        rowStreamUnpipe(rowStream, userStream);
+        userCanceled = true;
+        if (activeRequestStream) {
+          activeRequestStream.abort();
         }
-
-        // Remove rowKeys already read.
-        rowKeys = rowKeys.filter(rowKey =>
-          TableUtils.greaterThan(rowKey, lastRowKey as string)
-        );
-
-        // If there was a row limit in the original request and
-        // we've already read all the rows, end the stream and
-        // do not retry.
-        if (hasLimit && rowsLimit === rowsRead) {
-          userStream.end();
-          return;
+        originalEnd();
+        function isEncoding(
+          e: BufferEncoding | (() => void)
+        ): e is BufferEncoding {
+          return true;
         }
-        // If all the row keys and ranges are read, end the stream
-        // and do not retry.
-        if (rowKeys.length === 0 && ranges.length === 0) {
-          userStream.end();
-          return;
+        if (encodingOrCb) {
+          if (encodingOrCb && isEncoding(encodingOrCb)) {
+            return originalEnd(chunkOrCb, encodingOrCb, cb);
+          } else {
+            return originalEnd(chunkOrCb, encodingOrCb);
+          }
+        } else {
+          return originalEnd(chunkOrCb); // In practice, this code path is used.
```

Comment:
```suggestion
        if(!encodingOrCb){
          return originalEnd(chunkOrCb); 
        }
        if(!isEncoding(encodingOrCb)){
          return originalEnd(chunkOrCb, encodingOrCb);
        }
        return originalEnd(chunkOrCb, encodingOrCb, cb)
```

I left a comment about the code paths below, but if all of the paths have to stay, I simplified the logic with guard clauses. Thanks Alex Fenster for teaching me that term 😁 (see go/tott/651 for an example)

Reply:
I think your point above was correct about the fact that we can remove the other code paths.

Reply:
So I removed the other code paths instead.